rsocket/internal/RSocketConnectionManager.cpp

```diff
@@ -2,42 +2,71 @@
 
 #include "rsocket/internal/RSocketConnectionManager.h"
 
-#include <folly/io/async/EventBase.h>
-#include <folly/ScopeGuard.h>
 #include <folly/ExceptionWrapper.h>
+#include <folly/ScopeGuard.h>
+#include <folly/io/async/EventBase.h>
 
 #include "rsocket/RSocketConnectionEvents.h"
 #include "rsocket/statemachine/RSocketStateMachine.h"
 
 namespace rsocket {
 
 RSocketConnectionManager::~RSocketConnectionManager() {
-  // Asynchronously close all existing ReactiveSockets.  If there are none, then
-  // we can do an early exit.
-  VLOG(1) << "Destroying RSocketConnectionManager...";
-  auto scopeGuard = folly::makeGuard([]{ VLOG(1) << "Destroying RSocketConnectionManager... DONE"; });
+  VLOG(1) << "Started ~RSocketConnectionManager";
+  SCOPE_EXIT {
+    VLOG(1) << "Finished ~RSocketConnectionManager";
+  };
 
+  StateMachineMap map;
+
+  // Set up the baton and counter of connections yet to be closed.  Move all the
+  // connections out of the synchronized map so we don't block while keeping the
+  // map locked.
   {
     auto locked = sockets_.lock();
     if (locked->empty()) {
+      VLOG(2) << "No connections to close, early exit";
       return;
     }
 
-    shutdown_.emplace();
+    shutdownBaton_.emplace();
+    shutdownCounter_ = locked->size();
+
+    VLOG(2) << "Need to close " << shutdownCounter_.load() << " connections";
+
+    map.swap(*locked);
```

Comment:
It took some time to wrap my head around all the changes.  Have you considered just making a copy of the map (instead of clearing `locked`).  Then you wouldn't need the complexity in `removeConnection()`.

Reply:
I don't like making a copy of the map when I can do a move instead.

I don't think `removeConnection` is that complex.  I could make it even simpler if we had a Latch implementation in the `rsocket/` directory (I have one in `benchmarks/` but I meant it for testing only).